artifact/tar/platform/libarchive/wrapper.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <archive.h>
+
+#include <libarchive/wrapper.hpp>
+#include <common/log.hpp>
+
+namespace mender {
+namespace libarchive {
+namespace wrapper {
+
+// The reader_callback is invoked whenever the library requires raw bytes from
+// the archive. The read callback reads data into a buffer, sets the const void
+// **buffer argument to point to the available data, and returns a count of the
+// number of bytes available. LibArchive will invoke the read callback again
+// only after it has consumed this data. LibArchive imposes no constraints on
+// the size of the data blocks returned.
+//
+// - On EOF return 0.
+// - On error return -1.
+ssize_t reader_callback(archive *archive, void *in_reader_container, const void **buff) {
+	ReaderContainer *p_reader_container = reinterpret_cast<ReaderContainer *>(in_reader_container);
```

Comment:
Should only need `static_cast` here. Casting from `void` to anything is allowed.